Storage test enhancements

tg-msft opened the pull request:
- Change test recordings to not require a local config
- Add an initial CONTRIBUTING.md
- Add an `[Explicit]` test to garbage collect any resources in our test tenants

Please filter out all the `*.json` files when reviewing.  I had to re-record all of the tests to include the config.

tg-msft committed:
Storage test enhancements:

- Change test recordings to not require a local config
- Add an initial CONTRIBUTING.md
- Add a test to garbage collect any resources in our test tenants

pakrym reviewed (approved):
LGTM on test recording related stuff.

tg-msft committed:
Add Storage specific record matching and update test recordings

tg-msft committed:
A few testing fixes

- Change the diagnostic attribute name
- Use the same buffer size across platforms so recordings are in sync

pakrym commented on sdk/core/Azure.Core/tests/TestFramework/RecordMatcher.cs:
```diff
@@ -26,7 +26,11 @@ public RecordMatcher(RecordedTestSanitizer sanitizer)
             "User-Agent",
-            "Request-Id"
+            "Request-Id",
+            "If-Match",
```
I'm not sure about this, `If*` headers are something that clients generate based on inputs. What is the reason you would like to avoid matching them?

pakrym replied:
I guess because of equivalency check?

tg-msft replied:
They're always ETags or Dates.  I can move them to just the StorageRecordMatcher if you don't think they should generally be ignored.

tg-msft replied:
Separated `ExcludedHeaders` for matching and `VolatileHeaders` for ignoring unchanged values on re-record.

pakrym commented on sdk/core/Azure.Core/tests/TestFramework/RecordSession.cs:
```diff
@@ -93,7 +93,11 @@ public bool IsEquivalent(RecordSession session, RecordMatcher matcher)
-            return session.Variables.SequenceEqual(Variables) &&
+            // The DateTimeOffsetNow variable is updated any time it's used so
+            // we only care that both sessions use it or both sessions don't.
+            var now = TestRecording.DateTimeOffsetNowVariableKey;
```
Why not do the same thing ID reuse does where we get the value from the previous recording if available? Would avoid having this check.

tg-msft replied:
The service won't let us use DateTime values in the past so it always needs to be updated when we re-record.

pakrym commented on sdk/core/Azure.Core/tests/TestFramework/RecordMatcher.cs:
```diff
@@ -83,11 +88,28 @@ public virtual RecordEntry FindMatch(Request request, IList<RecordEntry> entries
 
-        public virtual bool IsEquivalentResponse(RecordEntry entry, RecordEntry otherEntry)
+        public virtual bool IsEquivalentRecord(RecordEntry entry, RecordEntry otherEntry) =>
+            IsEquivalentRequest(entry, otherEntry) &&
```
Does it matter that request it the same too?

tg-msft replied:
Yes - if I changed the URI but it generated an identical response, I want the recording to be serialized.  (I hit this specifically because I tried to change the URI to force a recording to serialize and was stumped when nothing happened.)

pakrym commented:
I wish I've added more tests for this feature :(

tg-msft committed:
Separate excluded headers from volatile headers in recording matches

and be a little more careful about waiting for progress events

tg-msft committed:
Fix rename outside of Storage

tg-msft committed:
Fixing #6806 to remove excess serialization whitespace

(which should also help recordings across platforms)

pakrym commented:
:shipit:
